### ISSUE TITLE
If i change the img_size, should i modify the anchors???

### ISSUE THREAD

**Issue by Fred199683** [ultralytics/yolov5#3116]
## ❔Question
The default anchors are belows(img_size=640):
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
 I think the anchors' size is related to the receptive field of the feature map. But if i change the img_size(640---->1280),so the objects are bigger, but the feature map's RF is not changed. So how to handle the anchors? To match the bigger objects, I think i should magnify the anchors, but this is not consistent with the RF.....
Hope anyone can help me with this! Thx

**Comment by glenn-jocher** [comment]
@Fred199683 👋 Hello! Thanks for asking about model anchors. YOLOv5 🚀 uses a new Ultralytics algorithm called **autoanchor** for anchor verification and generation before training starts.

Autoanchor will analyse your anchors against your dataset and training settings (like `--img-size`), and will adjust your anchors as necessary if it determines the original anchors are a poor fit, or if an **anchor count** was specified in your model.yaml rather than **anchor values**, i.e. 
```yaml
# Specify anchor count (per layer)
anchors: 3

# --OR-- Specify anchor values manually
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```

When generating new anchors, autoanchor first applies a kmeans function against your dataset labels (scaled to your training `--img-size`), and then uses these kmeans centroids as initial conditions for a Genetic Evolution (GE) algorithm. The GE algorithm will evolve all anchors for 1000 generations under default settings, using the CIoU loss (same one used during training) combined with Best Possible Recall (BPR) as its fitness function.

No action is required on your part to use autoanchor. If you would like to force manual anchors for any reason, you can skip autoanchor with the `--noautoanchor` flag:
```shell
python train.py --noautoanchor
```

**Comment by fcakyon** [comment]
@glenn-jocher what about inference?

**Comment by glenn-jocher** [comment]
@fcakyon for best results you should run inference at up to the same size you trained at.

**Comment by gohguodong** [comment]
Hi, if autoanchor is on during training. may i know where i can get the anchor values used for the trained model? i am asking because i need the anchors in the ONNX and tensor RT deployment.

**Comment by glenn-jocher** [comment]
@gohguodong YOLOv5 anchors are saved as [Detect()](https://github.com/ultralytics/yolov5/blob/fab5085674f7748dc16d7ca25afb225fa441bc9d/models/yolo.py#L22) layer attributes on model creation, and updated as necessary by autoanchor before training starts. Their exact location is here:
https://github.com/ultralytics/yolov5/blob/fab5085674f7748dc16d7ca25afb225fa441bc9d/models/yolo.py#L34-L35

You can examine the anchors of any trained YOLOv5 model like this:

### Input
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', autoshape=False)  # official model
model = torch.hub.load('ultralytics/yolov5', 'custom', 'path/to/best.pt',  autoshape=False)  # custom model

# Anchors
m = model.model[-1]  # Detect() layer
print(m.anchor_grid.squeeze())  # print anchors
```

### Output
```python
#           x     y
tensor([[[ 10.,  13.],
         [ 16.,  30.],
         [ 33.,  23.]],  # P3/8-small

        [[ 30.,  61.],
         [ 62.,  45.],
         [ 59., 119.]],  # P4/16-medium

        [[116.,  90.],
         [156., 198.],
         [373., 326.]]], dtype=torch.float16)  # P5/32-large
```

**Comment by youyuxiansen** [comment]
@glenn-jocher 
Is there any paper about the Ultralytics algorithm called "autoanchor"?

**Comment by sunbuhui** [comment]
@youyuxiansen 
Maybe you can check the code here by yourself.
https://github.com/ultralytics/yolov5/blob/407dc5008e47b1aad5ce69f0c91b4f1ec321dd7f/utils/autoanchor.py
